mitodl/bootcamp-ecommerce
# Use single-task and retry429 decorators for hubspot sync functions

#### What are the relevant tickets?
Closes #1392

#### What's this PR do?
Adds a new `single_task` decorator to prevent simultaneous API requests to Hubspot for the same object.  Also adds decorators to retry tasks on 429 exceptions (similar to 
https://github.com/mitodl/mitxonline/issues/1330)

#### How should this be manually tested?
You'll probably only get 429 errors if you have lots of users or deals to sync.  But assuming you do:

Set `MITOL_HUBSPOT_API_PRIVATE_TOKEN` to the same value as RC, set `MITOL_HUBSPOT_API_ID_PREFIX` to a value unique to you.

Set `HUBSPOT_MAX_CONCURRENT_TASKS=10` (or higher)

Run `docker-compose run web python manage.py sync_hubspot_to_db --users <create|update>`

Check the celery logs, if 429's are thrown (might not be) they should be retried until successful or 3 tries have been made.  Tasks should complete successfully either way.


In a python shell, run the following:
```python
from django.contrib.auth.models import User
user = User.objects.first()
from hubspot_sync.tasks import sync_contact_with_hubspot
for i in range(10):
    sync_contact_with_hubspot.delay(user.id)
```

Check the container logs, you'll likely see some of these.  Eventually the tasks should all succeed.
```
celery_1    | [2023-01-17 18:09:27,832: INFO/ForkPoolWorker-7] Task hubspot_sync.tasks.sync_contact_with_hubspot[c6157a63-8027-437b-a761-083f2f52f6c0] retry: Retry in 47s: BlockingIOError()
```
